staging/src/k8s.io/kubectl/pkg/util/completion/completion.go

```diff
@@ -94,6 +94,33 @@ func PodResourceNameCompletionFunc(f cmdutil.Factory) func(*cobra.Command, []str
 	}
 }
 
+// PodResourceNameAndContainerPortCompletionFunc Returns a completion function that completes, as a first argument:
+// 1- pod names that match the toComplete prefix
+// 2- resource types containing pods which match the toComplete prefix
+// and as a second argument the containers' ports within the specified pod.
+func PodResourceNameAndContainerPortCompletionFunc(f cmdutil.Factory) func(*cobra.Command, []string, string) ([]string, cobra.ShellCompDirective) {
+	return func(_ *cobra.Command, args []string, toComplete string) ([]string, cobra.ShellCompDirective) {
+		var comps []string
+		var resourceType string
+		directive := cobra.ShellCompDirectiveNoFileComp
+		if len(args) == 0 {
+			comps, directive = doPodResourceCompletion(f, toComplete)
+		} else if len(args) == 1 {
+			t := strings.Split(args[0], "/")
+			// if we specify directly the pod, then resource type is pod, otherwise it would be the resource type passed
+			// by the user
+			if len(t) == 1 {
+				resourceType = "pod"
+			} else {
+				resourceType = t[1]
+			}
+			resourceName := convertResourceNameToPodName(f, args[0])
```

Comment:
At first, I didn't see how this feature would ever work for deployments, statefulsets, etc because I didn't see anywhere that `CompGetContainersPorts()` was handling those types (the template would be different).

But I *think* what is going on is that `convertResourceNameToPodName()` is (by calling `polymorphichelpers.AttachablePodForObjectFn`) searching for and finding an underlying pod and returning that pod instead.

If that is the case, I think it should set resourceType = "pod" explicitly after the call to `convertResourceNameToPodName()` so that it is clear that we are now working with a pod.